extensions/amp-form/0.1/amp-form.js

```diff
@@ -107,16 +127,17 @@ export class AmpForm {
       return;
     }
 
-    const shouldValidate = !this.form_.hasAttribute('novalidate');
-    if (shouldValidate &&
-        this.form_.checkValidity && !this.form_.checkValidity()) {
-      e.preventDefault();
-      // TODO(#3776): Use .mutate method when it supports passing state.
-      this.vsync_.run({
-        measure: undefined,
-        mutate: reportValidity,
-      }, {form: this.form_});
-      return;
+    if (this.shouldValidate_) {
+      checkUserValidityOnSubmission_(this.form_);
+      if (this.form_.checkValidity && !this.form_.checkValidity()) {
```

Comment:
I can't really figure out a way to re-write this to only do `checkValidity` once without possibly passing a pre-calculated validity around the functions.

```javascript
function checkUserValidityOnSubmission_(form, opt_isCurrentlyValid) {
  ...
  checkUserValidity(form, opt_isCurrentlyValid);
}

function checkUserValidity(element, opt_isCurrentlyValid) {
  ...
  let isCurrentlyValid = opt_isCurrentlyValid;
  if (isCurrentlyValid === undefined) {
    isCurrentlyValid = element.checkValidity();
  }
  ...
}
```

Or changing the signature of `checkUserValidity` to return an object literal with multiple values, namely `{shouldPropagate, isValid}`.

Any other ideas?